Return undefined when query param is not provided (i.e. Zod optional params)

kylem123 opened the pull request:
Encountered issue when using optional or nullish props in Zod schema. Essentially, the get query param function will attempt to fetch these props, but return an empty array. This then causes the zod schema to error, as it is expecting string.optional for example, but it receives an array. This PR alters this behaviour to return undefined in the case that the array is empty.

ardatan commented:
Thanks for the PR!
Could you add a changeset for a release(using `yarn changeset`) and a test as well to prevent future regressions?

ardatan commented on packages/fets/src/createRouter.ts:
```diff
@@ -153,6 +153,9 @@ export function createRouterBase(
           {
             get(_, prop) {
               const allQueries = url.searchParams.getAll(prop.toString());
+              if (prop !== "then" && allQueries.length === 0) {
```
Why do we check `then`?

kylem123 commented on packages/fets/tests/router.spec.ts:
```diff
@@ -158,7 +159,32 @@ describe('Router', () => {
     const response = await router.fetch('https://foo.com/foo?cat=dog&foo=bar&foo=baz&missing=');
     const json = await response.json();
-    expect(json).toMatchObject({ cat: 'dog', foo: ['bar', 'baz'], missing: [] });
+    expect(json).toMatchObject({ cat: 'dog', foo: ['bar', 'baz'], missing: undefined });
```
Yes, good spot, will address that too!

kylem123 commented on packages/fets/tests/router.spec.ts:
```diff
@@ -158,7 +159,32 @@ describe('Router', () => {
-    expect(json).toMatchObject({ cat: 'dog', foo: ['bar', 'baz'], missing: [] });
+    expect(json).toMatchObject({ cat: 'dog', foo: ['bar', 'baz'], missing: undefined });
+  });
+  it('can handle missing Zod query params', async () => {
```
Have moved it to this file

ardatan commented on packages/fets/tests/router.spec.ts:
```diff
@@ -1,3 +1,4 @@
+import { z } from 'zod';
```
Could you remove this extra import?

kylem123 commented:
I believe I have addressed the issues you have mentioned. I have also added the result of the `yarn changeset`. Please let me know if anything else is needed :)

kylem123 commented:
Hi, wondered if you'd seen that this PR is ready to go. Would be good to be merged ASAP as we are currently having to run a yarn patch to handle this.

ardatan commented:
https://github.com/ardatan/feTS/pull/223/files#r1238646645

This is still needed because `missing=` shouldn't be parsed as `undefined` but an empty string;
You can test the behavior in any JS env;
```ts
const url = new URL('http://dummy.com?missing=');
console.log(url.searchParams.get('missing'))
```

kylem123 commented:
> https://github.com/ardatan/feTS/pull/223/files#r1238646645
> 
> Unfortunately not :( This is still needed because `missing=` shouldn't be parsed as `undefined` but an empty string; You can test the behavior in any JS env;
> 
> ```ts
> const url = new URL('http://dummy.com?missing=');
> console.log(url.searchParams.get('missing'))
> ```

Hi, I believe I have resolved this with my latest commit, so it should now have the correct behaviour.

ardatan commented:
Thanks @kylem123 ! I'll merge it after CI finishes!

kylem123 commented:
> Thanks @kylem123 ! I'll merge it after CI finishes!

Awesome, thank you!
